File: exercise/22/solution1.cpp
class Solution {
public:
    void generate(vector<string> &result, string now, int left, int right, int n) {
        if (right == n) {
            result.push_back(now);
            return;
        }
        if (right < left) {
            generate(result, now + ")", left, right + 1, n);
        }
        if (left < n) {
            generate(result, now + "(", left + 1, right, n);
        }
    }

    vector<string> generateParenthesis(int n) {
        vector<string> result;
        generate(result, "", 0, 0, n);
        return result;
    }
};

File: exercise/22/solution1.py
class Solution:
    def generate(self, result: List[str], left: int, right: int, remain: int, now: str):
        if left > right:
            self.generate(result, left, right + 1, remain, now + ')')
        if left < remain:
            self.generate(result, left + 1, right, remain, now + '(')
        if left == right and left == remain:
            result.append(now)

    def generateParenthesis(self, n: int) -> List[str]:
        result = list()
        self.generate(result, 0, 0, n, "")
        return result

